database/migrations/2024_06_01_173029_add_anonymous_questions_to_evaluation_form.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+use Illuminate\Database\Migrations\Migration;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+use Illuminate\Support\Facades\DB;
+
+return new class () extends Migration {
+    /**
+     * Run the migrations.
+     */
+    public function up(): void
+    {
+        Schema::create('answer_sheets', function (Blueprint $table) {
+            $table->id();
+            $table->unsignedSmallInteger('semester_id');
+            // $table->boolean('is_resident'); // but the BBs... let this be a question
+            $table->unsignedSmallInteger('year_of_acceptance');
+            // no timestamps (for better anonymity)
```

Comment:
Huh; I see... could this somehow be solved quickly? I have no good ideas for that; besides adding a random id for each answer sheet, long answer and answer sheet-option connection... and even then, it is probably not cryptographically secure:D

Reply:
Well, this has been a question for me for quite some time. And that is why I proposed to simply use an embedded Google Forms sheet.

The order in which records are created are probably stored regardless of the identifier so randomizing it is probably not enough.

In theory, it is possible to anonymously collect the answers. These are **not** a recommendations:

- we could create a record for each user when the question is created where it is indicated whether they have already voted. The record can be updated when they vote. The order on which these updates are created is probably not possible to retrieve afterwards if binary logging is disabled. (If binary logging is enabled, than it is trivial to restore.) Probably it would still be possible using forensic tools but I intentionally wrote about the nominal state of the schemas.
- currently, we add a new record when a user submits an answer. After this, we could create a random permutation of `user_id`s. It has the same limitations as I the previous solution. And it also increases processing time, and one would need to be sure to make this procedure atomic, consistent and isolated from other requests which are not trivial (not just throwing the code into a transaction as it can cause dead locks).

As always, a miscreant system administrator can simply wiretap the HTTP requests to access the information.

I would probably not implement any of my proposed solutions. They seem to be quite cumbersome and not easy to implement, and I just wanted to show that it **is** possible to get to a point where it is extremely difficult to restore who answered what. And also it only limits people with who have direct control over the server which is quite a small group.

This issue is easier to solve when data is aggregated during data collection.

Some resolutions:
- we simply accept that system administrators do see who answered what
- we do not store the fact whether a user has submitted an answer

Reply:
I think we should simply go with accepting that; it would be important for them to see that their questions do get stored and checked. Thank you for the ideas:)

Reply:
I think `// no timestamps (for better anonymity)` could be misleading. I would add `, the order in the database might break anonymity`, as it is highly likely, that no one would look back into this conversation.

Reply:
It is trivial that by using a centralized for web interface voting, the voters need to trust the system administrators (it is impossible to fix this - https://youtu.be/LkH2r-sNjQs ). But system administrators should know what kind of data should be handled really carefully.

Reply:
As I understand, the current implementation would only provide pseudonymouzation in case of long answers and would provide proper anonymization in cases where it is aggregated during the voting.

My main takeaway here is not the video's title I linked (that electronic voting is bad). But rather that we should be aware of our system's limitations.
That is why I asked what do we call an anonymous answer.

This is not necessarily a problem.

So we could change it to `// no timestamps (for better pseudonymity)`

Reply:
What about this?

```
// no timestamps (for better pseudonimity)
// note: the order of records might still break anonymity
```

Reply:
Yes, I like it